pd/src/validator_set.rs

```diff
@@ -0,0 +1,775 @@
+            // their signatures to get a deterministic ordering.
+            highest_seq_bucket.1.sort_by(|a, b| {
+                let a_sig = a.auth_sig.to_bytes();
+                let b_sig = b.auth_sig.to_bytes();
+                a_sig.cmp(&b_sig)
+            });
+
+            // Our pick will be the first definition in the bucket after sorting by signature.
+            resolved_validator_definitions.insert(ik.clone(), highest_seq_bucket.1[0].clone());
+        }
+
+        // Now that we have resolved all validator definitions, we can determine the validator
+        // changes that occurred in this block.
+        for (ik, def) in resolved_validator_definitions.iter() {
+            if self.validators().any(|v| v.borrow().identity_key == *ik) {
+                // If this is an existing validator, there will need to be a database UPDATE query.
+                // The existing state will be maintained but the validator configuration will change
+                // to the new definition.
+                // (TODO: ensure funding stream changes are properly accounted for).
+                let mut validator_info = self.validator_set.get_mut(ik).unwrap();
+
+                // Update the internal validator configuration
+                validator_info.validator = def.validator.clone();
+
+                // Add the validator to the block's updated validators list so an UPDATE query will be generated in
+                // `commit_block`.
+                self.updated_validators.push(validator_info.clone());
+            } else {
+                // Create the new validator's ValidatorInfo struct.
+                // The status will default to Inactive for new validators.
+                let new_validator = make_validator(def.clone());
+
+                // Add the validator to the internal validator set.
+                self.add_validator(new_validator.clone());
+
+                // Add the validator to the block's new validators list so an INSERT query will be generated in `commit_block`.
+                self.new_validators.push(new_validator);
+            }
+        }
+
+        // Set `self.tm_validator_updates` to the complete set of
+        // validators and voting power. This must be the last step performed,
+        // after all voting power calculations and validator state transitions have
+        // been completed.
+        //
+        // TODO: It could be more efficient to only return the power of
+        // updated validators.
+        self.tm_validator_updates = self
+            .validators_info()
+            .map(|v| {
+                let v = v.borrow();
+                // if the validator is non-Active, set their voting power as
+                // returned to Tendermint to 0. Only Active validators report
+                // voting power to Tendermint.
+                let power = if v.status.state == ValidatorState::Active {
+                    v.status.voting_power as u32
```

Comment:
I did this because `tendermint::vote::power::Power` only has certain `From` impls which don't include `u64`, but there _is_ a `TryFrom` for `u64` so I'll use that.